backend/de.metas.printing/de.metas.printing.base/src/main/java/de/metas/printing/process/C_Invoice_Candidate_EnqueueSelectionForInvoicingAndPDFConcatenating.java

```diff
@@ -0,0 +1,166 @@
+{
+	final IQueryBL queryBL = Services.get(IQueryBL.class);
+	//
+	// Services
+	private final IWorkPackageQueueFactory workPackageQueueFactory = Services.get(IWorkPackageQueueFactory.class);
+	private final IAsyncBatchBL asyncBatchBL = Services.get(IAsyncBatchBL.class);
+	// Parameters
+	private IInvoicingParams invoicingParams;
+
+	private int selectionCount = 0;
+
+	@Param(parameterName = I_C_Invoice_Candidate.COLUMNNAME_AD_Org_ID, mandatory = true)
+	private int p_AD_Org_ID;
+
+	@Param(parameterName = I_M_Product.COLUMNNAME_M_Product_Category_ID, mandatory = true)
+	private int p_M_Product_Category_ID;
+
+	@Override
+	@RunOutOfTrx
+	protected void prepare()
+	{
+		setShowProcessLogs(ShowProcessLogs.OnError);
+
+		final IParams params = getParameterAsIParams();
+		this.invoicingParams = new InvoicingParams(params);
+
+		selectionCount = createSelection();
+		if (selectionCount <= 0)
+		{
+			final Properties ctx = getCtx();
+			throw new AdempiereException(msgBL.getMsg(ctx, IInvoiceCandidateEnqueuer.MSG_INVOICE_GENERATE_NO_CANDIDATES_SELECTED_0P));
+		}
+
+	}
+
+	private AsyncBatchId createAsyncBatch()
+	{
+		return asyncBatchBL.newAsyncBatch(C_Async_Batch_InternalName_InvoiceCandidate_Processing);
+	}
+
+	@Override
+	protected String doIt() throws Exception
+	{
+		final AsyncBatchId asyncBatchId = createAsyncBatch();
+		final I_C_Async_Batch asyncBatch = asyncBatchBL.getAsyncBatchById(asyncBatchId);
+		asyncBatchBL.setPInstance_IDAndSave(asyncBatch, getPinstanceId());
+
+		final IWorkPackageQueue queue = workPackageQueueFactory.getQueueForEnqueuing(getCtx(), InvoiceEnqueueingWorkpackageProcessor.class);
+		queue.newBlock()
+				.setContext(getCtx())
+				.newWorkpackage()
+				.setC_Async_Batch(asyncBatchBL.getAsyncBatchById(asyncBatchId))
+				.parameters(invoicingParams.asMap())
+				.build();
+
+		return MSG_OK;
+	}
+
+	private int createSelection()
+	{
+		final IQueryBuilder<I_C_Invoice_Candidate> queryBuilder = createICQueryBuilder();
+
+		//
+		// Create selection and return how many items were added
+		final PInstanceId adPInstanceId = getPinstanceId();
+		Check.assumeNotNull(adPInstanceId, "adPInstanceId is not null");
+
+		DB.deleteT_Selection(adPInstanceId, ITrx.TRXNAME_ThreadInherited);
+
+		final int selectionCount = queryBuilder
+				.create()
+				.setRequiredAccess(Access.READ)
+				.createSelection(adPInstanceId);
+
+		return selectionCount;
+	}
+
+	private IQueryBuilder<I_C_Invoice_Candidate> createICQueryBuilder()
+	{
+		final IQuery<I_M_Product_Category> subQuery_ProductcategFilter = queryBL
+				.createQueryBuilder(I_M_Product_Category.class)
+				.addEqualsFilter(I_M_Product_Category.COLUMNNAME_M_Product_Category_ID, p_M_Product_Category_ID)
+				.setJoinOr()
+				.addEqualsFilter(I_M_Product_Category.COLUMNNAME_M_Product_Category_Parent_ID, p_M_Product_Category_ID)
+				.create();
+
+		final IQuery<I_M_Product> subQuery_Product = queryBL
```

Comment:
i suggest to refactor and introduce `IQuery<I_M_Product> IProductBL.queryByProductCategory(...)` method.